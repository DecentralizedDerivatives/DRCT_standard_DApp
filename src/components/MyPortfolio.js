import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Collapse } from 'reactstrap';
import MyPositions from './MyPositions';
import MyTransactions from './MyTransactions';
import ContractDetails from './ContractDetails';
import CreateContract from "./CreateContract";
import {
  getUserAccount,
  getUserPositions,
  getUserTransactions
} from '../actions/userActions';
import requireConnection from './requireConnection';

// Use named export for unconnected component for testing
export class MyPortfolio extends Component {
  constructor() {
    super();

    this.state = {
      detailsOpen: false,
      formOpen: false,
    };
  }

  async componentDidMount() {
    await this.props.getUserAccount();
    this.props.getUserPositions(this.props.userAccount);
    this.props.getUserTransactions(this.props.userAccount);
  }

  // async componentDidUpdate() {
  //   await this.props.getUserPositions(this.props.userAccount);
  //   await this.props.getUserTransactions(this.props.userAccount);
  // }

  handleRowClick = async e => {
    e.preventDefault();

    let addressEl = e.target.getElementsByClassName('link__token-address')[0];

    if (typeof addressEl !== 'undefined') {
      const token_address = addressEl.getAttribute('data-token-address');

      await this.props.setSelectedToken(token_address);

      this.openContractDetails(e.target, token_address);
    }
  };

  openContractDetails = async (link, token_address = false) => {
    await this.props.getContractDetails(link);

    this.setState({
      detailsOpen: true
    });
  };

  closeContractDetails = () => {
    this.setState({
      detailsOpen: false
    });
  };
  handleCreateContract = () => {
    this.setState({
      formOpen: !this.state.formOpen
    });
  };
  renderCreateContract = () =>{
    return this.state.formOpen?(
      <CreateContract close={this.closeCreateContract} />
    ):(
      null
    )
  }
  closeCreateContract = () =>{
    this.setState({
      formOpen: false
    });
  }
  render() {
    return (
      <div id="portfolio">
        <MyPositions onRowClick={this.handleRowClick.bind(this)} />
        <MyTransactions onRowClick={this.handleRowClick.bind(this)} />
        <Collapse isOpen={this.state.detailsOpen}>
          <ContractDetails onClick={this.closeContractDetails.bind(this)} />
        </Collapse>
        {this.renderCreateContract()}
        <div className="create-contract-btn" onClick={this.handleCreateContract}>Create Contract</div>
      </div>
    );
  }
}

MyPortfolio.propTypes = {
  getUserAccount: PropTypes.func.isRequired,
  getUserPositions: PropTypes.func.isRequired,
  getUserTransactions: PropTypes.func.isRequired,
  orderId: PropTypes.string,
  userAccount: PropTypes.string.isRequired
};

const mapStateToProps = state => ({
  userAccount: state.user.userAccount
});

export default connect(
  mapStateToProps,
  { getUserAccount, getUserPositions, getUserTransactions }
)(requireConnection(MyPortfolio));