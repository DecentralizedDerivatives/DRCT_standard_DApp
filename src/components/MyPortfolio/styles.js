export default theme => ({
  item: {
    width: '1000px',
    marginTop: '25px',
  },
  container: {
    marginBottom: '25px',
  },
});
