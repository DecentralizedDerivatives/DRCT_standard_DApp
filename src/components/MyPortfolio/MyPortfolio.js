import React, { Component } from 'react';
import PropTypes from 'prop-types';
import withStyles from 'material-ui/styles/withStyles';
import Grid from 'material-ui/Grid';
import styles from './styles';
import Table from '../Table';
import ContractDetails from '../ContractDetails';
import { Factory, Exchange, web3, DRCT } from '../../ethereum';

class MyPortfolio extends Component {
  constructor() {
    super();
    this.state = {
      previousActive: '',
      active: '',
      open: false,
      myPositions: [['loading...', 'loading...', 'loading...']],
      myTransactions: [['loading...', 'loading...']],
      myAccount: '',
      selectedTokenAddress:'',
      contractAddress: "",
      contractDuration: "",
      contractMultiplier: "",
      oracleAddress: "",
    };
  }
  componentDidMount() {
    web3.eth.getAccounts((error, accounts) => {
      this.setState({ myAccount: accounts[0] });
    });    
    //Getting contract details one time when its parent gets mounted
    this.getContractDetails();
    this.getMyPositions().then(result => {
      this.setState({ myPositions: result });
    });
    this.getmyTransactions();
  }

  onClickRow = link => {
    let addressEl =  link.currentTarget.getElementsByClassName("token-address-link")[0];
    if(typeof addressEl !== "undefined"){
      this.openContractDetails(link,addressEl.getAttribute("data-token-address"));
    }
  };

  openContractDetails = (newActive,token_address=false) => {
    if(token_address){
      this.setState({ active:newActive, open: true, previousActive: this.state.active,selectedTokenAddress:token_address });
    }else{
      this.setState({ active:newActive, open: true, previousActive: this.state.active});
    }
  };
  getContractDetails = async () => {
    const factory = await Factory.at("0x15bd4d9dd2dfc5e01801be8ed17392d8404f9642");
    let response, error;
    try {
      response = await factory.getVariables();
    } catch (err) {
      error = err;
    }
    if (error) {
      console.error("Error getting contract details",error);
      return;
    }
    this.setState({
      contractAddress: response[0],
      contractDuration: response[1].c[0],
      contractMultiplier: response[2].c[0],
      oracleAddress: response[3],
    });
  };
  closeContractDetails = () => {
    this.setState({
      open: false,
      active: this.state.previousActive,
    });
  };

  getMyPositions = async () => {
    const factory = await Factory.at("0x15bd4d9dd2dfc5e01801be8ed17392d8404f9642");
    const _allrows = [];
    const openDates = [];
    const numDates = await factory.getDateCount();
    for (let i = 0; i < numDates; i++) {
      const startDates = (await factory.startDates.call(i)).c[0];
      const _token_addresses = await factory.getTokens(startDates);
      let _date = new Date(startDates * 1000);
      _date = (_date.getUTCMonth() + 1) + '/' + (_date.getUTCDate()) + '/' + _date.getUTCFullYear();
      openDates.push(_date);
      for (let j = 0; j < _token_addresses.length; j++) {
        let drct = await DRCT.at(_token_addresses[j]);
        let _balance = await drct.balanceOf(this.state.myAccount);
        if (_balance.c[0] > 0) {
          _allrows.push({
            address:_token_addresses[j],
            balance:_balance.c[0].toString(),
            date:_date.toString(),
            symbol:"ETH/USD", /*CURRENTLY USING STATIC SYMBOL NEED TO FIX*/
            contractAddress:this.state.contractAddress,
            contractDuration:this.state.contractDuration,
            contractMultiplier:this.state.contractMultiplier,
            oracleAddress:this.state.oracleAddress,
          });
        }
      }
    }
    if (_allrows.length === 0) _allrows.push({address:'No Current Positions', balance:'...', date:'...',symbol:"..."});
    return _allrows;
  };

  getmyTransactions = async () => {
    const exchange = await Exchange.deployed();
    const factory = await Factory.at("0x15bd4d9dd2dfc5e01801be8ed17392d8404f9642")
    let drct;
    let _trades = [];
    var titles = ['ContractCreation']; //Add other ATS when redeployed
    var ats = [
      {
        //Sale: await exchange.Sale({}, {fromBlock:0, toBlock: 'latest'}),
        //OrderPlaced: await exchange.OrderPlaced({}, {fromBlock:0, toBlock: 'latest'}),
        //OrderRemoved: await exchange.OrderRemoved({}, {fromBlock:0, toBlock: 'latest'}),
        ContractCreation: await factory.ContractCreation(
          {},
          { fromBlock: 0, toBlock: 'latest' }
        ),
        //Transfer: await drct.Transfer({}, {fromBlock:0, toBlock: 'latest'}),
        //Approval: await drct.Approval({}, {fromBlock:0, toBlock: 'latest'})
      },
    ];

    for (let i = 0; i < titles.length; i++) {
      let transferEvent = await factory.ContractCreation(
        {},
        { fromBlock: 0, toBlock: 'latest' }
      );
      await transferEvent.get((error, logs) => {
        for (let j = logs.length - 1; j >= Math.max(logs.length - 10, 0); j--) {
          if (
            logs[i].args['_sender'].toUpperCase() ==
            this.state.myAccount.toUpperCase()
          ) {
            _trades.push([titles[i], logs[j].transactionHash]);
            this.setState({ myTransactions: _trades });
          }
        }
        if (_trades.length === 0) {
          _trades = [['No Recent Events', '...']];
        }
      });
    }
  };

  render() {
    const { classes } = this.props;

    return (
      <Grid
        container
        className={classes.container}
        direction="row"
        alignItems="center"
        align="center"
        justify="center"
      >
        <Grid item className={classes.item}>
          <Table
            titles={['Asset', 'Balance', 'Start Date']}
            rows={this.state.myPositions}
            tableWidth="950px"
            clickFunction={this.onClickRow}
          />
        </Grid>
        <Grid item className={classes.item}>
          <Table
            titles={['My Transactions', 'Transaction Hash']}
            rows={this.state.myTransactions}
            tokenInfo={this.state.tokenInfo}
            tableWidth="950px"
            cellHeight="15px"
            fontSize="12px"
            clickFunction={this.onClickRow}
          />
        </Grid>
        <ContractDetails
          open={this.state.open}
          toggle={this.closeContractDetails}
          contractAddress={this.state.contractAddress}
          contractDuration={this.state.contractDuration}
          contractMultiplier={this.state.contractMultiplier}
          oracleAddress={this.state.oracleAddress}
          tokenAddress={this.state.selectedTokenAddress}
        />
      </Grid>
    );
  }
}

MyPortfolio.propTypes = {
  classes: PropTypes.object.isRequired,
};
export default withStyles(styles)(MyPortfolio);
