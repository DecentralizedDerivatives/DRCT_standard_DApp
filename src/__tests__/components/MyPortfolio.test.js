import MyPortfolio from '../../components/MyPortfolio';

describe('<MyPortfolio />', () => {
  it('renders the component', () => {
    const wrapper = shallow(<MyPortfolio store={initFixtureStore()} />)
      .dive()
      .dive()
      .dive();

    expect(wrapper).toMatchSnapshot();
  });
});
