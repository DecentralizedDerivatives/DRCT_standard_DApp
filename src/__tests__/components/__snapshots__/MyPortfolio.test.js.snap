// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MyPortfolio /> renders the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MyPortfolio
    dispatch={[Function]}
    getUserAccount={[Function]}
    getUserPositions={[Function]}
    getUserTransactions={[Function]}
    metamask={true}
    network={4}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
          "unsubscribe": [Function],
        },
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      }
    }
    userAccount="0xd8e09fbab17f30e1C2b2Ae85A26EA1F3F31267ef"
    verified={true}
    whiteListed={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Connect(MyPositions) />,
        <Connect(MyTransactions) />,
        null,
        <div
          className="btn btn-theme create-contract-btn"
          onClick={[Function]}
        >
          Create Contract
        </div>,
        null,
        <div
          id="cashout-button"
        >
          <button
            className="btn btn-theme create-contract-btn"
            onClick={[Function]}
          >
            Cash Out
          </button>
        </div>,
      ],
      "id": "portfolio",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Create Contract",
          "className": "btn btn-theme create-contract-btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Create Contract",
        "type": "div",
      },
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <button
            className="btn btn-theme create-contract-btn"
            onClick={[Function]}
          >
            Cash Out
          </button>,
          "id": "cashout-button",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Cash Out",
            "className": "btn btn-theme create-contract-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Cash Out",
          "type": "button",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Connect(MyPositions) />,
          <Connect(MyTransactions) />,
          null,
          <div
            className="btn btn-theme create-contract-btn"
            onClick={[Function]}
          >
            Create Contract
          </div>,
          null,
          <div
            id="cashout-button"
          >
            <button
              className="btn btn-theme create-contract-btn"
              onClick={[Function]}
            >
              Cash Out
            </button>
          </div>,
        ],
        "id": "portfolio",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Create Contract",
            "className": "btn btn-theme create-contract-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Create Contract",
          "type": "div",
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <button
              className="btn btn-theme create-contract-btn"
              onClick={[Function]}
            >
              Cash Out
            </button>,
            "id": "cashout-button",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Cash Out",
              "className": "btn btn-theme create-contract-btn",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Cash Out",
            "type": "button",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
